nixos/modules/services/networking/unbound.nix

```diff
@@ -6,41 +6,113 @@ let
-
-  forward =
-    optionalString (any isLocalAddress cfg.forwardAddresses) ''
-      do-not-query-localhost: no
-    '' +
-    optionalString (cfg.forwardAddresses != []) ''
-      forward-zone:
-        name: .
-    '' +
-    concatMapStringsSep "\n" (x: "    forward-addr: ${x}") cfg.forwardAddresses;
-
-  rootTrustAnchorFile = "${stateDir}/root.key";
-
-  trustAnchor = optionalString cfg.enableRootTrustAnchor
-    "auto-trust-anchor-file: ${rootTrustAnchorFile}";
-
-  confFile = pkgs.writeText "unbound.conf" ''
-    server:
-      directory: "${stateDir}"
-      username: unbound
-      chroot: "${stateDir}"
-      pidfile: ""
-      ${interfaces}
-      ${access}
-      ${trustAnchor}
-    ${cfg.extraConfig}
-    ${forward}
-  '';
+  settingsSubmodule = let
+    validConfigTypes = with types; either int (either str (either bool float));
+  in {
+    options = {
+      server = mkOption {
+        type = with types; attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes)));
+        default = {
+          directory = cfg.stateDir;
+          username = cfg.user;
+          chroot = cfg.stateDir;
+          pidfile = ''""'';
+          interface = [ "127.0.0.1" ] ++ (optional config.networking.enableIPv6 "::1");
+          access-control = [ "127.0.0.0/24 allow" ];
+          auto-trust-anchor-file = if cfg.enableRootTrustAnchor then rootTrustAnchorFile else null;
+        };
+        description = "server: clause configuration";
+      };
+      remote-control = mkOption {
+        type = with types; attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes)));
+        default = {
+          control-enable = false;
+          control-interface = [ "127.0.0.1" ] ++ (optional config.networking.enableIPv6 "::1");
+          server-key-file = "${cfg.stateDir}/unbound_server.key";
+          server-cert-file = "${cfg.stateDir}/unbound_server.pem";
+          control-key-file = "${cfg.stateDir}/unbound_control.key";
+          control-cert-file = "${cfg.stateDir}/unbound_control.pem";
+        };
+        description = "remote-control: clause configuration";
+      };
+      stub-zone = mkOption {
+        type = with types; listOf (attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes))));
+        default = [ ];
+        description = "stub-zone: clause configuration";
+      };
+      forward-zone = mkOption {
+        type = with types; listOf (attrsOf (nullOr (either validConfigTypes (listOf validConfigTypes))));
+        default = [ ];
+        example = literalExample ''
+          [
+            {
+              name = ".";
+              forward-addr = "1.1.1.1@853#cloudflare-dns.com";
+            }
+            {
+              name = "example.org.";
+              forward-addr = [
+                "1.1.1.1@853#cloudflare-dns.com"
+                "1.0.0.1@853#cloudflare-dns.com"
+              ];
+            }
+          ];
+        '';
+        description = "forward-zone: clauses configuration";
```

Comment:
Please link upstream documentation on possible values.

Reply:
Is a reference to a manpage engouh?

Reply:
Yes :+1: